### PR TITLE
modules

### LINKED ISSUE #23: Switching to a module system
What's up with having no module system?

At least, that's what it seems to me.

### LINKED ISSUE #81: check installation scripts
from the goodness squad meeting, it seems that they are problematic.

### PR DESCRIPTION
**Pull request by amir-arad**
TL;DR : the codebase changed from using global namespaces to using explicitly imported es6 modules. 

fixes #81
fixes #23

these are the major differences:
1. **each file is its own namespace and module.** files with internal namespaces (or modules) were split to several files. so that file `foo.ts` with internal namespace `bar` will now be `foo.ts` and `foo.bar.ts`. This is not a technical requirement of the new "order", but rather a normalization of the code that helped the transition. In some cases (the repeating builder+descriptor sub-namespaces) it may be better to revert it once things are stable (TBD)
2. **the `application.ts` file:** 
 - in addition to the internal namespaces (`bus`, `location`, `terminal`), it spawned 2 additional files in order to overcome a mean dependency spaghetti: 
   - `constants.ts` : version and events names. 
   - `kickstart.ts` : the application boot logic.   
 - also, the exported variables `context` and `mainView` are no longer exported. they now hide behind exported getter-setter functions. This was part of the things I tried while attacking the spaghetti. I think it's safe to revert this specific change if desired.
3. **[webpack](https://webpack.github.io/) is in the house.** the project now uses webpack both as:
 - a final build step using `npm run build` , it bundles the transpilation results and injects a`<script>` tag to the `html` file for that bundle, basically bundling `kickstart` with all its dependencies.  
 - a dev environment instead of the previous static server (same port, same npm command, but now it bundles the modules whenever one of them changes)
also, the index.html file moved to the `assets` folder and was stripped of all script tags.
webpack does not currently handle source maps. it's possible to do so, but then the bundle gets huge, and some problems get harder to debug in the browser. TBD.
4. **[`.editorconfig`](http://editorconfig.org/) file was added to the root of the project.** its contents are not my personal opinion, but rather what I saw to be the de-facto standard of this project. I used auto-formatting a **lot**, and this helped me to avoid git noise.
5. last but not least: **there is now the new concept of modules API.** having no global namespace means that modules loaded dynamically can't just reach to every part of the client they want. So, `kickstart.ts` also saves a designated namespace 'fugazi' on the window object containing all the hooks the modules should need. The file `scripts/bin/app/modules.api.d.ts` describes these hooks (the `fugazi` property on `window` and its type) and supplies additional types that modules may need. This is an early attempt at formalizing the API. I don't like this approach, but it was the right one for the transition as it introduced the least changes.

### CODE REVIEW

**Comment by amir-arad** [comment]
fixes #81 
fixes #23

**Comment by nitzantomer** [comment]
All of the json files are changes because json files are now using 2 spaces as indentation instead of tabs.  
It's not the preference (for me at least), but more than that it adds changed files to a big PR

**Comment by nitzantomer** [comment]
the `modules/restheart` directory can be deleted.
it will be easier to see those files as deleted instead of changes in those files (that are not really relevant because the entire dir isn't needed anymore)

File: modules/restheart/scripts/src/db.converters.ts [hunk 1]
```diff
@@ -1,28 +1,20 @@
...
+(function (): void {
+	fugazi.loaded({
 		name: "restheart.databases",
 		converters: {
 			RawDatabases2Databases: {
 				title: "converts RawDatabases to Databases",
 				input: "RawDatabases",
 				output: "Databases",
-				converter: function(response: { _embedded: { _id: string }[] }): string[] {
+				converter: function (response: { _embedded: { _id: string }[] }): string[] {
```
Comment (nitzantomer) [inline modules/restheart/scripts/src/db.converters.ts]: can this space addition be removed?
it also creates unimportant changes in a big PR

File: modules/restheart/scripts/tsconfig.json [hunk 2]
```diff
@@ -1,9 +1,9 @@
 {
-	"compilerOptions": {
-		"target": "ES6",
-		"removeComments": true,
-		"sourceMap": true,
-		"outDir": "bin",
-		"rootDir": "src"
-	}
+  "compilerOptions": {
```
Comment (nitzantomer) [inline modules/restheart/scripts/tsconfig.json]: again, tabs changed to two spaces in json files

File: modules/scripts/src/base/collections.ts [hunk 3]
```diff
@@ -1,23 +1,18 @@
-/// <reference path="../../../../scripts/bin/components/components.d.ts" />
-/// <reference path="../../../../scripts/bin/components/modules.d.ts" />
+import {BoundConstraintValidator, Descriptor, Map} from "../../../../scripts/bin/app/modules.api";
```
Comment (nitzantomer) [inline modules/scripts/src/base/collections.ts]: maybe change `Map` to `FugaziMap` (or something like that) so there won't be a mix-up with the native `Map` object

File: modules/scripts/src/base/commands.ts [hunk 4]
```diff
@@ -1,12 +1,7 @@
-/// <reference path="../../../../scripts/bin/app/application.d.ts" />
-/// <reference path="../../../../scripts/bin/components/registry.d.ts" />
-/// <reference path="../../../../scripts/bin/components/components.d.ts" />
+import {Component, Descriptor, Map, ModuleContext} from "../../../../scripts/bin/app/modules.api";
```
Comment (nitzantomer) [inline modules/scripts/src/base/commands.ts]: Same here, `FugaziMap` (or what not)

File: modules/scripts/src/base/net.http.ts [hunk 5]
```diff
@@ -1,21 +1,15 @@
-/// <reference path="../../../../scripts/bin/core/net.d.ts" />
+import {Descriptor, HttpMethod, Map, ModuleContext, RequestProperties} from "../../../../scripts/bin/app/modules.api";
```
Comment (nitzantomer) [inline modules/scripts/src/base/net.http.ts]: `FugaziMap`

File: modules/scripts/tsconfig.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
```
Comment (nitzantomer) [inline modules/scripts/tsconfig.json]: json file indentation changes only.
this file and a few right afterwards

File: scripts/src/app/application.ts [hunk 7]
```diff
@@ -1,217 +1,77 @@
...
-		type: components.types.Type,
-		value: any
+import * as viewMain from "../view/main";
+import * as utils from "../core/utils";
+import * as registry from "../components/registry";
+import * as converters from "../components/converters";
+import * as types from "../components/types";
+
+
+var context: ApplicationContext;
```
Comment (nitzantomer) [inline scripts/src/app/application.ts]: let

File: scripts/src/app/application.ts [hunk 8]
```diff
@@ -1,217 +1,77 @@
...
+export function getContext() {
+	return context;
+}
+export function setContext(newContext: ApplicationContext) {
+	context = newContext;
+}
+
+var mainView: viewMain.MainView;
```
Comment (nitzantomer) [inline scripts/src/app/application.ts]: let

File: modules/restheart/scripts/src/db.converters.ts (hunk 1 above)
Comment (amir-arad) [reply modules/restheart/scripts/src/db.converters.ts]: (that specific file was removed as it was in `modules/restheart`, I'm addressing the general issue)
regarding additional whitespaces, I hear you. agreed, it adds noise that reduces the ability to review.
I've used the IDE's auto-format. tracking down these spaces by hand can be a lot of manual work. if there's a de-facto style rule we should keep, can we express it in a machine-friendly format so that the computer will take care of it?

File: modules/scripts/src/base/collections.ts (hunk 3 above)
Comment (amir-arad) [reply modules/scripts/src/base/collections.ts]: agreed and fixed.
also I have a lot of phase 2 thoughts (regarding the polyfills, the kisckstart, analytics and in general, re-usability),  but I don't see these issues as blockers.

File: modules/scripts/src/base/commands.ts (hunk 4 above)
Comment (amir-arad) [reply modules/scripts/src/base/commands.ts]: done

File: modules/scripts/src/base/net.http.ts (hunk 5 above)
Comment (amir-arad) [reply modules/scripts/src/base/net.http.ts]: done

File: modules/scripts/tsconfig.json (hunk 6 above)
Comment (amir-arad) [reply modules/scripts/tsconfig.json]: done

File: scripts/src/app/application.ts (hunk 8 above)
Comment (amir-arad) [reply scripts/src/app/application.ts]: done.

there are a lot of `var`s roaming free in the code. some of them should be `const` and some `let`.

phase 2 IMO.
